src/oauth/oauth2.js

```diff
@@ -106,9 +109,18 @@ export default class OAuth2 {
       exchangeTokenUrl = this.providerConfig.url
     }
 
-    return this.$http.post(exchangeTokenUrl, payload, {
-      withCredentials: this.options.withCredentials
-    })
+    let config = {
+        withCredentials: this.options.withCredentials
+    };
+
+    if (this.providerConfig.confidentialClient === true) {
+      config.auth = {
+        username: this.providerConfig.clientId,
+        password: this.providerConfig.clientSecret
```

Comment:
Oh, this is a problem. As i said in your other PR #141 , this is a major security issue and i'm not going to merge this into my library. This is probably the reason why Satellizer (which my library is built upon) doesn't even support password oauth flow - because it requires client secret.

Primary task of this library is to provide JWT authentication based on OAuth 1/2 protocols.

Reply:
>  As i said in your other PR #141 ,

I missed this - can you please reference your comment? THX

Reply:
> this is a major security issue and i'm not going to merge this into my library.

Well - client secrets are distributed quite publicly in some cases and yes they can no longer be considered that secret ;-)

Would it be valid to give the user of this library the chance to use it in an environment where clients have to authenticate via id+secret?

Reply:
It's the only comment in #141 PR :)

Reply:
Somebody manipulating the matrix ..... ?

![screenshot from 2018-10-08 22-36-54](https://user-images.githubusercontent.com/1005065/46632371-b76a1000-cb4a-11e8-88f9-b158f52047d4.png)

Reply:
Sorry, just realised that i forgot to submit review. Now you'll be able to see it.

Reply:
But, it is the same comment as this one above.

Reply:
Hello, it is necessary to pass the secretKey, for example, in the mautic API to obtain the access token. I need it for this and with the changes of @DeepDiver I can get the token.

See the petition examples for Mautic.

https://developer.mautic.org/#oauth-2

Reply:
That is true, but those are all server-side examples because it is a huge security risk in exposing `client_secret` in SPA application. I would suggest you try to find some other solution for your problem: for example, collect username & password in SPA and send it to your server so that the server can do the authentication and just return received token to SPA app.

I am not familiar with what you are developing on your project, but what you are trying to do with this library seems like a bad idea.

Reply:
All providers supported by this library have a way of ensuring a person's identity before they generate an access token. For example, when you want to login using Facebook provider, you have to log in to Facebook using your FB email and password. After your successful login, you'll get access token. But, when `grant_type === 'code'`, you need a server-side implementation which will exchange code for access token - and in that request, you need client secret. This is why that request is kept on server-side, instead of client-side.

Reply:
Even if I have the clientID and the secretKey, if I do not have a username and password to access the application they do not help me. In my case, when I make the request for the token it asks me to authenticate in the application (mautic) and then launches the method `exchangeForToken`. Therefore, I do not mind storing these keys in the client since without a valid username and password I will not be able to do anything.

Reply:
Well, implement it as you wish, but I'm not adding support for client secret in this library. That will allow developers to use it in a very insecure way - not something I want to encourage. If you really need this implementation, feel free to fork repo and do whatever you need.

Reply:
> All providers supported by this library have a way of ensuring a person's identity before they generate an access token. For example, when you want to login using Facebook provider, you have to log in to Facebook using your FB email and password. After your successful login, you'll get access token.

But there is also another request performed which exchanges the code for the access token. This is the code section we are talking about. What grant_type is e.g. FB using here?



> Even if I have the clientID and the secretKey, if I do not have a username and password to access the application they do not help me. In my case, when I make the request for the token it asks me to authenticate in the application (mautic) and then launches the method `exchangeForToken`. Therefore, I do not mind storing these keys in the client since without a valid username and password I will not be able to do anything.

Full ack on this. ClientID and ClientSecret alone help nothing in terms of getting access. They are basically used for determining if a physical client is allowed to talk to your system. They are not about granting a specific user to access the system.

But maybe I'm mixing up things ......

Reply:
@dgrubelic don't get my comments wrong - this is not about convincing you - this is about getting an understanding of the oauth mess in the world ;-)

Reply:
@dgrubelic If I understand correctly the usage scenario of your lib is purely the implicit grant - right? Taking account the comment from https://oauth.net/2/grant-types/implicit/ the advise is to go for authorization code with client secret.

Reply:
No, it is not. If you see example and server.js file (https://github.com/dgrubelic/vue-authenticate/blob/master/example/server.js), you can see that you can do full OAuth 2 flow using any of the supported (or custom) OAuth providers. There you can use client secret in exchange code for token request.

Reply:
This is true if you have your own server implementation - but this does not follow the oauth spec.
If you run the token exchange you are in an authorization grant. And this would require the changes in this PR.

Reply:
Anyway - as explained below I use implicit grant and this lib works as a charm. THX :beers:

Reply:
> but this does not follow the oauth spec

I think that this actually is not entirely true. Simply because, if you go to this link https://tools.ietf.org/html/rfc6749#section-1.3.1 and see the last change date (October 2012), when OAuth 2 protocol spec last change was written, SPA's were not this popular and basically every web applications developed had a server-side implementation - they actually were classic server-side applications that rendered entire HTML on the server. So, technically, we could consider that authors of OAuth specs actually considered that you will have closed and secured server environment where client secret wasn't exposed as it could be in SPA's. Even though they speak about "client", in my opinion, it is meant as:
- client: an application that sends authentication requests (at that time server-side application)
- authentication server: a server that contains sensitive user data and actually does authentication exchange and provides the client with auth token.

Only in the last few years, when SPA's became popular, term "client" in OAuth protocol specs can actually be both secure server-side application AND insecure (regarding client secret) single page application. I guess this is causing some confusion.

With this, i think that we can close this argument and PR as well.

Reply:
For the sake of completness: https://www.oauth.com/oauth2-servers/single-page-apps/

It talks about implicit flow as well as authorization code flow WITHOUT secret. So you are totally right. Only the grant type is missing in the request to fully support this.

Anyway - let's close this PR for now.

Thanks a lot! :beers: